File: packages/core/skills/extracting-requirements/scripts/chunk_spec.mjs
import { readdirSync, readFileSync, statSync } from "node:fs";
import { basename, join, resolve, sep } from "node:path";
import { pathToFileURL } from "node:url";

export function estimateTokens(text) {
  const words = text.trim() ? text.trim().split(/\s+/).length : 0;
  return Math.trunc(words * 1.3);
}

export function splitByHeading(content, level) {
  const pattern = new RegExp(`^${"#".repeat(level)} (.+)$`, "gm");
  const matches = [...content.matchAll(pattern)];

  if (matches.length === 0) return [{ heading: null, content }];

  const sections = [];
  if (matches[0].index > 0) {
    const preamble = content.slice(0, matches[0].index).trim();
    if (preamble) sections.push({ heading: "(preamble)", content: preamble });
  }

  for (const [index, match] of matches.entries()) {
    const start = match.index;
    const end = matches[index + 1]?.index ?? content.length;
    sections.push({
      heading: match[1].trim(),
      content: content.slice(start, end).trim(),
    });
  }

  return sections;
}

function lineCount(content) {
  return (content.match(/\n/g) ?? []).length;
}

export function findLineRange(fullContent, sectionContent) {
  const position = fullContent.indexOf(sectionContent.slice(0, 80));
  if (position === -1) return [1, lineCount(fullContent) + 1];

  const startLine = lineCount(fullContent.slice(0, position)) + 1;
  return [startLine, startLine + lineCount(sectionContent)];
}

export function chunkFile(path, maxTokens) {
  const content = readFileSync(path, "utf8");
  const tokens = estimateTokens(content);

  if (tokens <= maxTokens) {
    return [
      {
        source_file: path,
        heading: null,
        start_line: 1,
        end_line: lineCount(content) + 1,
        content,
        estimated_tokens: tokens,
      },
    ];
  }

  const sections = splitByHeading(content, 2);
  const chunks = [];
  for (const section of sections) {
    const sectionTokens = estimateTokens(section.content);
    const [startLine, endLine] = findLineRange(content, section.content);

    if (sectionTokens <= maxTokens) {
      chunks.push({
        source_file: path,
        heading: section.heading,
        start_line: startLine,
        end_line: endLine,
        content: section.content,
        estimated_tokens: sectionTokens,
      });
      continue;
    }

    for (const subsection of splitByHeading(section.content, 3)) {
      const subsectionTokens = estimateTokens(subsection.content);
      const [subStart, subEnd] = findLineRange(content, subsection.content);
      const heading =
        subsection.heading && subsection.heading !== "(preamble)"
          ? `${section.heading} > ${subsection.heading}`
          : section.heading;
      chunks.push({
        source_file: path,
        heading,
        start_line: subStart,
        end_line: subEnd,
        content: subsection.content,
        estimated_tokens: subsectionTokens,
      });
    }
  }

  return chunks;
}

function markdownFiles(path) {
  const files = [];
  const entries = readdirSync(path, { withFileTypes: true }).sort((left, right) => {
    if (left.name < right.name) return -1;
    if (left.name > right.name) return 1;
    return 0;
  });
  for (const entry of entries) {
    const candidate = join(path, entry.name);
    if (entry.isDirectory()) files.push(...markdownFiles(candidate));
    else if (entry.isFile() && entry.name.endsWith(".md")) files.push(candidate);
  }
  return files;
}

export function chunkPath(path, maxTokens) {
  if (statSync(path).isFile()) return chunkFile(path, maxTokens);
  if (statSync(path).isDirectory()) return markdownFiles(path).flatMap((file) => chunkFile(file, maxTokens));
  return [];
}

function usage(program) {
  return `usage: ${program} [-h] [--max-tokens MAX_TOKENS] path\n`;
}

function help(program) {
  return [
    usage(program).trimEnd(),
    "",
    "Chunk spec files for extraction",
    "",
    "positional arguments:",
    "  path                  File or directory to chunk",
    "",
    "options:",
    "  -h, --help            show this help message and exit",
    "  --max-tokens MAX_TOKENS",
    "                        Max tokens per chunk (default 4000)",
    "",
  ].join("\n");
}

function parseTokenCount(value) {
  if (!/^[-+]?\d+$/.test(value)) return null;
  return Number(value);
}

function pathParts(path, separatorPattern) {
  return path.split(separatorPattern).filter((part) => part && part !== ".");
}

function normalizeWindowsPathSpelling(path) {
  const windowsPath = path.replaceAll("/", "\\");
  const unc = windowsPath.match(/^\\\\([^\\]+)\\([^\\]+)(?:\\|$)/);
  if (unc) {
    const root = `\\\\${unc[1]}\\${unc[2]}\\`;
    const parts = pathParts(windowsPath.slice(unc[0].length), /\\+/);
    return parts.length > 0 ? `${root}${parts.join("\\")}` : root;
  }

  const drive = windowsPath.match(/^([A-Za-z]:)(\\?)/);
  if (drive) {
    const rooted = drive[2] === "\\";
    const parts = pathParts(windowsPath.slice(drive[1].length + (rooted ? 1 : 0)), /\\+/);
    if (rooted) return parts.length > 0 ? `${drive[1]}\\${parts.join("\\")}` : `${drive[1]}\\`;
    return parts.length > 0 ? `${drive[1]}${parts.join("\\")}` : drive[1];
  }

  const rooted = windowsPath.startsWith("\\");
  const parts = pathParts(windowsPath.slice(rooted ? 1 : 0), /\\+/);
  if (rooted) return parts.length > 0 ? `\\${parts.join("\\")}` : "\\";
  return parts.length > 0 ? parts.join("\\") : ".";
}

function normalizePathSpelling(path) {
  if (sep === "\\") return normalizeWindowsPathSpelling(path);

  const doubleSlashRoot = path.startsWith("//") && !path.startsWith("///");
  const root = doubleSlashRoot ? "//" : path.startsWith("/") ? "/" : "";
  const parts = pathParts(path, /\/+/);
  if (root) return parts.length > 0 ? `${root}${parts.join("/")}` : root;
  return parts.length > 0 ? parts.join("/") : ".";
}

function parseArgs(args) {
  let path;
  let maxTokens = 4000;
  const unrecognized = [];

  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (arg === "-h" || arg === "--help") return { help: true };
    if (arg === "--max-tokens" || arg.startsWith("--max-tokens=")) {
      const value = arg === "--max-tokens" ? args[++index] : arg.slice("--max-tokens=".length);
      if (value === undefined) return { error: "argument --max-tokens: expected one argument" };
      const parsed = parseTokenCount(value);
      if (arg === "--max-tokens" && value.startsWith("-") && parsed === null) {
        return { error: "argument --max-tokens: expected one argument" };
      }
      if (parsed === null) return { error: `argument --max-tokens: invalid int value: '${value}'` };
      maxTokens = parsed;
    } else if (!path && !arg.startsWith("-")) {
      path = arg;
    } else {
      unrecognized.push(arg);
    }
  }

  if (!path) return { error: "the following arguments are required: path" };
  if (unrecognized.length > 0) return { error: `unrecognized arguments: ${unrecognized.join(" ")}` };
  return { path, maxTokens };
}

export async function main(args) {
  const program = basename(process.argv[1] ?? "chunk_spec.mjs");
  const options = parseArgs(args);
  if (options.help) {
    process.stdout.write(help(program));
    return 0;
  }
  if (options.error) {
    process.stderr.write(`${usage(program)}${program}: error: ${options.error}\n`);
    return 2;
  }

  const path = normalizePathSpelling(options.path);

  try {
    statSync(path);
  } catch {
    process.stderr.write(`error: path not found: ${path}\n`);
    return 2;
  }

  process.stdout.write(`${JSON.stringify(chunkPath(path, options.maxTokens), null, 2)}\n`);
  return 0;
}

const isDirect = process.argv[1] && pathToFileURL(resolve(process.argv[1])).href === import.meta.url;
if (isDirect) process.exitCode = await main(process.argv.slice(2));

File: plugins/moe/.claude-plugin/skills/extracting-requirements/scripts/chunk_spec.py
#!/usr/bin/env python3
"""Chunk spec files into sections for parallel extraction.

Usage: chunk_spec.py <path> [--max-tokens 4000]

If <path> is a file, chunks that single file.
If <path> is a directory, chunks all .md files recursively.

Output: JSON array of chunks to stdout. Each chunk has:
  source_file, heading, start_line, end_line, content, estimated_tokens

Chunking strategy:
  Files < max_tokens: single chunk (whole file)
  Files >= max_tokens: split by ## headings
  Sections still over max_tokens: split by ### headings
"""
import argparse
import json
import re
import sys
from pathlib import Path


def estimate_tokens(text: str) -> int:
    """Estimate token count. Conservative: 1 word ~ 1.3 tokens."""
    return int(len(text.split()) * 1.3)


def split_by_heading(content: str, level: int) -> list[dict]:
    """Split markdown by heading level. Returns list of {heading, content}."""
    prefix = "#" * level
    pattern = re.compile(rf"^{prefix} (.+)$", re.MULTILINE)
    matches = list(pattern.finditer(content))

    if not matches:
        return [{"heading": None, "content": content}]

    sections: list[dict] = []

    # Preamble before first heading
    if matches[0].start() > 0:
        preamble = content[: matches[0].start()].strip()
        if preamble:
            sections.append({"heading": "(preamble)", "content": preamble})

    for idx, match in enumerate(matches):
        start = match.start()
        end = matches[idx + 1].start() if idx + 1 < len(matches) else len(content)
        sections.append({
            "heading": match.group(1).strip(),
            "content": content[start:end].strip(),
        })

    return sections


def find_line_range(full_content: str, section_content: str) -> tuple[int, int]:
    """Find the start and end line numbers of section_content within full_content."""
    pos = full_content.find(section_content[:80])
    if pos == -1:
        return 1, full_content.count("\n") + 1
    start_line = full_content[:pos].count("\n") + 1
    end_line = start_line + section_content.count("\n")
    return start_line, end_line


def chunk_file(path: Path, max_tokens: int) -> list[dict]:
    """Chunk a single file into sections."""
    content = path.read_text()
    tokens = estimate_tokens(content)

    if tokens <= max_tokens:
        return [{
            "source_file": str(path),
            "heading": None,
            "start_line": 1,
            "end_line": content.count("\n") + 1,
            "content": content,
            "estimated_tokens": tokens,
        }]

    # Split by ## headings
    sections = split_by_heading(content, level=2)
    chunks: list[dict] = []

    for section in sections:
        sec_tokens = estimate_tokens(section["content"])
        start_line, end_line = find_line_range(content, section["content"])

        if sec_tokens <= max_tokens:
            chunks.append({
                "source_file": str(path),
                "heading": section["heading"],
                "start_line": start_line,
                "end_line": end_line,
                "content": section["content"],
                "estimated_tokens": sec_tokens,
            })
        else:
            # Sub-split by ### headings
            subsections = split_by_heading(section["content"], level=3)
            for subsec in subsections:
                sub_tokens = estimate_tokens(subsec["content"])
                sub_start, sub_end = find_line_range(content, subsec["content"])
                heading = section["heading"]
                if subsec["heading"] and subsec["heading"] != "(preamble)":
                    heading = f"{heading} > {subsec['heading']}"
                chunks.append({
                    "source_file": str(path),
                    "heading": heading,
                    "start_line": sub_start,
                    "end_line": sub_end,
                    "content": subsec["content"],
                    "estimated_tokens": sub_tokens,
                })

    return chunks


def chunk_path(path: Path, max_tokens: int) -> list[dict]:
    """Chunk a file or directory."""
    if path.is_file():
        return chunk_file(path, max_tokens)
    if path.is_dir():
        chunks: list[dict] = []
        for md_file in sorted(path.rglob("*.md")):
            chunks.extend(chunk_file(md_file, max_tokens))
        return chunks
    return []


def main() -> int:
    parser = argparse.ArgumentParser(description="Chunk spec files for extraction")
    parser.add_argument("path", help="File or directory to chunk")
    parser.add_argument("--max-tokens", type=int, default=4000,
                        help="Max tokens per chunk (default 4000)")
    args = parser.parse_args()

    path = Path(args.path)
    if not path.exists():
        print(f"error: path not found: {path}", file=sys.stderr)
        return 2

    chunks = chunk_path(path, args.max_tokens)
    json.dump(chunks, sys.stdout, indent=2)
    print()
    return 0


if __name__ == "__main__":
    sys.exit(main())
